cpp/src/strings/split/split.cu

```diff
@@ -385,6 +386,348 @@ struct whitespace_rsplit_tokenizer_fn
     }
 };
 
+// TODO: copied from cuDF PR 3685 (should include instead of duplicating once
+// PR 3685 is merged)
+#if 1
+template <typename T>
+__device__ inline T round_up_pow2(T number_to_round, T modulus) {
+    return (number_to_round + (modulus - 1)) & -modulus;
+}
+
+// align all column size allocations to this boundary so that all output column buffers
+// start at that alignment.
+static constexpr size_type split_align = 64;
+#endif
+
+/**
+ * @brief Compute the number of tokens, the total byte sizes of the tokens, and
+ * required memory size for the `idx'th` string element of `d_strings`.
+ */
+template <bool forward>
+struct token_reader_fn {
+  column_device_view const d_strings;  // strings to split
+  string_view const d_delimiter;  // delimiter for split
+  size_type const max_tokens = std::numeric_limits<size_type>::max();
+  bool const has_validity = false;
+
+  // returns a tuple of token count, sum of token sizes in bytes, and required
+  // memory block size
+  __device__ thrust::tuple<size_type, size_type, size_type>
+  operator()(size_type idx) const {
```

Comment:
Consider refactoring `operator()` into separate calls to several `__device__` member functions for readability.

Reply:
Thank you and I will try to refactor these four operator() functions to improve readability.

Reply:
Done.